template/data/src/main/java/co/nimblehq/template/data/service/Exceptions.kt

```diff
@@ -0,0 +1,12 @@
+package co.nimblehq.template.data.service
```

Comment:
Any reason why we didn't declare these exception classes at the domain level? 🤔 
Or do we assume that `UseCase` doesn't care what is being thrown and will let the `ViewModel` handle it? 😄

Reply:
@ryan-conway Hmm true. Currently, the `ApiException` depends on `ErrorResponse` and `HttpException` from the `data` module. It can not be moved into `domain` module for usecase to access 💭 

I would like to look for more ideas @nimblehq/android-chapter, should we move it to `domain` for usecase to access? If yes, I think we will need
- Create a new Error model in `domain` and map `ErrorResponse` to `Error` model.
- Get rid of the base `HttpException`, the `ApiException` will contain Error model & other other fields such as response's `code` or `message`.

Reply:
@luongvo since `:data` depends on `:domain`, I think moving the exceptions to `:domain` is fine. They can be used by both `:app` and `:data` components.

Reply:
@luongvo I agree, `domain` should have its own error models independent of any library or framework. For example, we may have something like this in `domain` module
```kotlin
class ApiException(
    val httpCode: Int,
    val apiMessage: String?
): Exception()
```
And the mapping will remain in `data` module
```kotlin
private fun Throwable.mapError(): Throwable {
    return when (this) {
        is UnknownHostException, 
        is InterruptedIOException -> NoConnectivityException
        is HttpException -> ApiException(code(), getErrorBody(this))
        else -> this
    }
}
```

Reply:
@AVI5HEK @ryan-conway Fixed in https://github.com/nimblehq/android-templates/pull/311/commits/fd17e4a8d0fdf0a966bb6ff94168f4203917d2f3